src/vcf/fasta.cpp

```diff
@@ -0,0 +1,131 @@
+#include "util/stream_utils.hpp"
+#include "util/string_utils.hpp"
+#include "vcf/compression.hpp"
+#include "vcf/fasta.hpp"
+#include "vcf/string_constants.hpp"
+
+std::shared_ptr<std::istream> create_gunzipped_stream(std::shared_ptr<std::istream> stream) {
+  std::shared_ptr<std::stringstream> ss(new std::stringstream());
+  try {
+    boost::iostreams::filtering_istream decompressedStream;
+    ebi::vcf::create_uncompressed_stream(*stream, ebi::vcf::GZ, decompressedStream);
+    boost::iostreams::copy(decompressedStream, *ss);
+  } catch (std::exception const &ex) {
+    BOOST_LOG_TRIVIAL(error) << ex.what();
+  }
+
+  return std::static_pointer_cast<std::istream>(ss);
+}
+
+std::shared_ptr<std::istream> create_decompressed_stream(std::shared_ptr<std::istream> stream) {
+  std::vector<char> line;
+  ebi::vcf::get_magic_num(*stream, line);
+  std::string file_extension = ebi::vcf::get_compression("", line);
+
+  if ( file_extension == ebi::vcf::GZ ) {
+    return create_gunzipped_stream(stream);
+  }
+
+  return stream;
+}
+
+std::shared_ptr<std::istream> create_input_stream(const std::string &path) {
+  if ( !ebi::util::is_remote_url(path) ) {
+    std::shared_ptr<std::ifstream> fstream(new std::ifstream());
+    ebi::util::open_local(*fstream, path, std::ifstream::binary);
+    return create_decompressed_stream(std::static_pointer_cast<std::istream>(fstream));
+  }
+
+  std::shared_ptr<std::stringstream> sstream(new std::stringstream());
+  ebi::util::open_remote(*sstream, path);
+  return create_decompressed_stream(std::static_pointer_cast<std::istream>(sstream));
+}
+
+ebi::vcf::fasta::FileBasedFasta::FileBasedFasta(const std::string& fasta_path, const std::string& fasta_index_path)
+{
+  BOOST_LOG_TRIVIAL(info) << "Reading from input FASTA file...";
+  mFastaInput = create_input_stream(fasta_path);
+
+  BOOST_LOG_TRIVIAL(info) << "Reading from input FASTA index file...";
+  mFastaIndex = bioio::read_fasta_index(*(create_input_stream(fasta_index_path)));
+}
+
+std::string
+ebi::vcf::fasta::FileBasedFasta::sequence(const std::string& contig, const size_t pos, const size_t length) {
+  return bioio::read_fasta_contig(*mFastaInput, mFastaIndex.at(contig), pos, length);
+}
+
+size_t
+ebi::vcf::fasta::FileBasedFasta::count(const std::string &contig) const {
+  return mFastaIndex.count(contig);
+}
+
+ebi::vcf::fasta::RemoteContig::RemoteContig()
+        : mCurl(new ebi::util::curl::Easy())
+{}
+
+std::string
+ebi::vcf::fasta::RemoteContig::sequence(const std::string& contig, const size_t pos, const size_t length) {
+  if ( mContig.count(contig) == 0 ) {
+    mContig[contig] = "";
+
+    // This contig is not downloaded, try download it from ENA.
+    if (contig.size() < 6) { // don't trust short contig names.
+      return "";
+    }
+
+    // Download this contig
+    std::string url = ebi::vcf::ENA_API_URL + contig + "&" + ebi::vcf::ENA_API_FASTA_FORMAT;
+    std::stringstream ss;
+    ebi::util::open_remote(ss, url);
```

Comment:
so this downloads the whole contig from ENA into memory? I could be ok with this for now, to have the basic workings, but I think we can optimise this to download chunks, and free the chunks after they no longer have the requested region, as the VCFs should be sorted

Reply:
what about downloading it into a file on hard-drive instead of downloading into a memory. this would also help in caching that file for later use. in case we need the same chunk in the next run.

Reply:
As mentioned above, the sequence will be written to a tmp file which later will be removed. If we want to cache those reference sequences for later use, there should be limit or a curated list to work with.